/**:
  ros__parameters:
    # Topics
    # ubran honkong
    pointCloudTopic: "/livox/lidar"               # Point cloud data
    imuTopic: "/imu/data_livox"                         # IMU data # <!-- liorf_localization_yjz_lucky_boy -->

    # # mulran
    # pointCloudTopic: "/os1_points"               # Point cloud data
    # imuTopic: "/imu/data_raw"                         # IMU data # <!-- liorf_localization_yjz_lucky_boy -->
    odomTopic: "odometry/imu"                   # IMU pre-preintegration odometry, same frequency as IMU
    gpsTopic: "/fix"                   # GPS odometry topic from navsat, see module_navsat.launch file

    # Frames
    lidarFrame: "livox_link"
    baselinkFrame: "base_link"
    odometryFrame: "liorf_odom"
    mapFrame: "map"

    # GPS Settings
    useImuHeadingInitialization: true           # if using GPS data, set to "true"
    useGpsElevation: true                      # if GPS elevation is bad, set to "false"
    gpsCovThreshold: 0.0002                        # m^2, threshold for using GPS data
    poseCovThreshold: 0.000000000001                      # m^2, threshold for using GPS data
    
    # Export settings
    savePCD: false                              # https://github.com/TixiaoShan/LIO-SAM/issues/3
    savePCDDirectory: "/data/"        # in your home folder, starts and ends with "/". Warning: the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation

    # Sensor Settings
    sensor: livox                            # lidar sensor type, 'velodyne' or 'ouster' or 'livox' or 'robosense'
    N_SCAN: 6                                  # number of lidar channel (i.e., Velodyne/Ouster: 16, 32, 64, 128, Livox Horizon: 6)
    Horizon_SCAN: 4000                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048, Livox Horizon: 4000)
    downsampleRate: 1                           # default: 1. Downsample your data if too many points(line). i.e., 16 = 64 / 4, 16 = 16 / 1
    point_filter_num: 2                         # default: 3. Downsample your data if too many points(point). e.g., 16: 1, 32: 5, 64: 8
    lidarMinRange: 0.7                          # default: 1.0, minimum lidar range to be used
    lidarMaxRange: 100.0                       # default: 1000.0, maximum lidar range to be used

    # IMU Settings
    imuType: 1                                  # 0: 6-axis  1: 9-axis
    imuRate: 200.0
    # hongkong ubran
    imuAccNoise: 3.9939570888238808e-03
    imuGyrNoise: 1.5636343949698187e-03
    imuAccBiasN: 6.4356659353532566e-04
    imuGyrBiasN: 3.5640318696367613e-04

    # # # MULRAN Datasets
    # imuAccNoise: 0.009939570888238808e-03
    # imuGyrNoise: 0.005636343949698187e-03
    # imuAccBiasN: 0.64356659353532566e-03
    # imuGyrBiasN: 0.35640318696367613e-03

    imuGravity: 9.80511
    imuRPYWeight: 0.01
    
    # # hongkong ubran
    # Extrinsics: T_lb (lidar -> imu)
    extrinsicTrans: [0.011, 0.023, -0.044]
    extrinsicRot: [1.0, 0.0, 0.0,
                    0.0, 1.0, 0.0,
                    0.0, 0.0, 1.0]

    # This parameter is set only when the 9-axis IMU is used, but it must be a high-precision IMU. e.g. MTI-680
    extrinsicRPY: [1.0, 0.0, 0.0,
                    0.0, 1.0, 0.0,
                    0.0, 0.0, 1.0]

    # # # mulran
    # # Extrinsics: T_lb (lidar -> imu)
    # extrinsicTrans: [0.0, 0.0, 0.0]
    # extrinsicRot: [-1, 0, 0,
    #                 0, -1, 0,
    #                 0, 0, 1]

    # # This parameter is set only when the 9-axis IMU is used, but it must be a high-precision IMU. e.g. MTI-680
    # extrinsicRPY: [-1, 0, 0,
    #                 0, -1, 0,
    #                 0, 0, 1]

    # voxel filter paprams
    mappingSurfLeafSize: 0.3                       # default: 0.4 - outdoor, 0.2 - indoor

    # robot motion constraint (in case you are using a 2D robot)
    z_tollerance: 1000.0                            # meters
    rotation_tollerance: 1000.0                     # radians

    # CPU Params
    numberOfCores: 8                              # number of cores for mapping optimization
    mappingProcessInterval: 0.2                  # seconds, regulate mapping frequency

    # Surrounding map
    surroundingkeyframeAddingDistThreshold: 1.0   # meters, regulate keyframe adding threshold
    surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold
    surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
    surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)
    surroundingKeyframeMapLeafSize: 0.4           # downsample local map point cloud

    # Loop closure
    loopClosureEnableFlag: true
    loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
    surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
    historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
    historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
    historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
    loopClosureICPSurfLeafSize: 0.5               # downsample icp point cloud  
    historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

    # Visualization
    globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
    globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
    globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density

    mappingGpsDatumLatitude: 45.53929754059958
    mappingGpsDatumLongitude: -122.92470385245429
    mappingGpsDatumAltitude: 39.0
    mappingGpsCloudTimeOffset: 0.0
    mappingGpsCovariance: 0.01
    mappingGpsAddingDist: 0.3
    
    mappingLmConvergenceTrans: 0.07
    mappingLmConvergenceRot: 0.07
    mappingProcessingTimeoutMs: 200
    maxNumOptimizationIterations: 24